delay-tasks/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+	pub enum Event<T: Config> {
+		DelayedTaskAdded {
+			id: Nonce,
+			task: T::Task,
+			execute_block: BlockNumberFor<T>,
+		},
+		DelayedTaskExecuted {
+			id: Nonce,
+			result: DispatchResult,
+		},
+		DelayedTaskReDelayed {
+			id: Nonce,
+			execute_block: BlockNumberFor<T>,
+		},
+		DelayedTaskTryExecuteFailed {
+			id: Nonce,
+			error: DispatchError,
+		},
+		DelayedTaskCanceled {
+			id: Nonce,
+		},
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// `on_initialize` to return the weight used in `on_finalize`.
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			Weight::zero()
+		}
+
+		fn on_finalize(now: BlockNumberFor<T>) {
+			Self::_on_finalize(now);
+		}
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_delayed_task_id)]
+	pub type NextDelayedTaskId<T: Config> = StorageValue<_, Nonce, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn delayed_tasks)]
+	pub type DelayedTasks<T: Config> = StorageMap<_, Twox64Concat, Nonce, (T::Task, BlockNumberFor<T>), OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn delayed_task_queue)]
+	pub type DelayedTaskQueue<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, BlockNumberFor<T>, Twox64Concat, Nonce, (), OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::zero())]
+		pub fn reset_execute_block(
+			origin: OriginFor<T>,
+			id: Nonce,
+			when: DispatchTime<BlockNumberFor<T>>,
+		) -> DispatchResult {
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			DelayedTasks::<T>::try_mutate_exists(id, |maybe_task| -> DispatchResult {
+				let (_, execute_block) = maybe_task.as_mut().ok_or(Error::<T>::InvalidId)?;
+
+				let now = frame_system::Pallet::<T>::block_number();
+				let new_execute_block = match when {
+					DispatchTime::At(x) => x,
+					DispatchTime::After(x) => x.checked_add(&now).ok_or(ArithmeticError::Overflow)?,
+				};
+				ensure!(new_execute_block > now, Error::<T>::InvalidDelayBlock);
+
+				DelayedTaskQueue::<T>::remove(*execute_block, id);
+				DelayedTaskQueue::<T>::insert(new_execute_block, id, ());
+				*execute_block = new_execute_block;
+
+				Self::deposit_event(Event::<T>::DelayedTaskReDelayed {
+					id,
+					execute_block: new_execute_block,
+				});
+				Ok(())
+			})?;
+
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::zero())]
+		pub fn cancel_delayed_task(origin: OriginFor<T>, id: Nonce) -> DispatchResult {
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			let (delay_task, execute_block) = DelayedTasks::<T>::take(id).ok_or(Error::<T>::InvalidId)?;
+			delay_task.on_cancel()?;
+			DelayedTaskQueue::<T>::remove(execute_block, id);
+
+			Self::deposit_event(Event::<T>::DelayedTaskCanceled { id });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn _on_finalize(now: BlockNumberFor<T>) {
```

Comment:
should do this with `on_idle` to be safer and we will know how many tasks to execute based on remaining weights